File: python/day17.py
def list_neigh_3d(r, c, z):
    return {
        (r - i, c - j, z - k)
        for i in range(-1, 2)
        for j in range(-1, 2)
        for k in range(-1, 2)
    } - {(r, c, z)}


def list_neigh_4d(r, c, z, w):
    return {
        (r - i, c - j, z - k, w - l)
        for i in range(-1, 2)
        for j in range(-1, 2)
        for k in range(-1, 2)
        for l in range(-1, 2)
    } - {(r, c, z, w)}


def step(state, list_neigh_fn):
    new_state = {i for i in state}
    extended_state = {n for p in state for n in list_neigh_fn(*p)} | state

    for p in extended_state:
        ncount = sum(n in state for n in list_neigh_fn(*p))
        if p in state and (ncount < 2 or 3 < ncount):
            new_state.remove(p)
        if p not in state and ncount == 3:
            new_state.add(p)
    return new_state


inp = """
....#...
.#..###.
.#.#.###
.#....#.
...#.#.#
#.......
##....#.
.##..#.#
""".strip()

state = {
    (i, j, 0)
    for i, line in enumerate(inp.split("\n"))
    for j, v in enumerate(line)
    if v == "#"
}

for i in range(6):
    state = step(state, list_neigh_3d)

print(f"3D count: {len(state)}")


inp = """
....#...
.#..###.
.#.#.###
.#....#.
...#.#.#
#.......
##....#.
.##..#.#
""".strip()
state = {
    (i, j, 0, 0)
    for i, line in enumerate(inp.split("\n"))
    for j, v in enumerate(line)
    if v == "#"
}
for i in range(6):
    state = step(state, list_neigh_4d)
print(f"4D count: {len(state)}")

File: src/day17.rs
use std::collections::HashSet;
use std::fs::read_to_string;
use std::path::Path;

fn list_neighbors((r, c, z): (isize, isize, isize)) -> Vec<(isize, isize, isize)> {
    vec![
        (r - 1, c - 1, z - 1),
        (r, c - 1, z - 1),
        (r + 1, c - 1, z - 1),
        (r - 1, c, z - 1),
        (r, c, z - 1),
        (r + 1, c, z - 1),
        (r - 1, c + 1, z - 1),
        (r, c + 1, z - 1),
        (r + 1, c + 1, z - 1),
        (r - 1, c - 1, z),
        (r, c - 1, z),
        (r + 1, c - 1, z),
        (r - 1, c, z),
        (r + 1, c, z),
        (r - 1, c + 1, z),
        (r, c + 1, z),
        (r + 1, c + 1, z),
        (r - 1, c - 1, z + 1),
        (r, c - 1, z + 1),
        (r + 1, c - 1, z + 1),
        (r - 1, c, z + 1),
        (r, c, z + 1),
        (r + 1, c, z + 1),
        (r - 1, c + 1, z + 1),
        (r, c + 1, z + 1),
        (r + 1, c + 1, z + 1),
    ]
}

fn count_active_neighbors(
    state: &HashSet<(isize, isize, isize)>,
    pos: (isize, isize, isize),
) -> usize {
    list_neighbors(pos)
        .iter()
        .filter(|pos| state.contains(pos))
        .count()
}

enum Action {
    Activate((isize, isize, isize)),
    Deactivate((isize, isize, isize)),
}

fn get_deltas(state: &HashSet<(isize, isize, isize)>) -> Vec<Action> {
    state
        .iter()
        .flat_map(|pos| list_neighbors(*pos))
        .collect::<HashSet<_>>()
        .iter()
        .copied()
        .filter_map(|pos| {
            let ncount = count_active_neighbors(&state, pos);
            let is_active = state.contains(&pos);
            match is_active {
                true if ncount < 2 => Some(Action::Deactivate(pos)),
                true if 3 < ncount => Some(Action::Deactivate(pos)),
                false if ncount == 3 => Some(Action::Activate(pos)),
                _ => None,
            }
        })
        .collect::<Vec<_>>()
}

fn apply_deltas(
    mut state: HashSet<(isize, isize, isize)>,
    deltas: Vec<Action>,
) -> HashSet<(isize, isize, isize)> {
    deltas.iter().for_each(|a| {
        match *a {
            Action::Activate(pos) => state.insert(pos),
            Action::Deactivate(pos) => state.remove(&pos),
        };
    });
    state
}

pub fn day17(path: &Path) {
    let mut state = HashSet::new();
    read_to_string(path)
        .unwrap()
        .split("\n")
        .enumerate()
        .flat_map(|(irow, row)| {
            row.chars()
                .enumerate()
                .filter_map(move |(icol, val)| match val {
                    '.' => None,
                    '#' => Some((irow as isize, icol as isize)),
                    _ => panic!(),
                })
        })
        .for_each(|(irow, icol)| {
            state.insert((irow, icol, 0));
        });
    for _ in 0..6 {
        let deltas = get_deltas(&state);
        state = apply_deltas(state, deltas);
    }
    println!("{:?}", state);
}
